src/Abuse/Adapters/Redis.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+namespace Utopia\Abuse\Adapters;
+
+use Utopia\Abuse\Adapter;
+use Redis as Client;
+
+class Redis extends Adapter
+{
+    public const NAMESPACE = 'abuse';
+
+    /**
+     * @var Client
+     */
+    protected Client $redis;
+
+    /**
+     * @var int
+     */
+    protected int $time;
+
+    /**
+     * @var int
+     */
+    protected int $limit = 0;
+
+    /**
+     * @var int|null
+     */
+    protected ?int $count = null;
+
+    public function __construct(string $key, int $limit, int $seconds, Client $redis)
+    {
+        $this->redis = $redis;
+        $this->key = $key;
+        $time = (int) \date('U', (int) (\floor(\time() / $seconds)) * $seconds); // todo: any good Idea without time()?
+        $this->time = $time;
+        $this->limit = $limit;
+    }
+
+    /**
+     * Undocumented function
+     *
+     * @param string $key
+     * @param int $datetime
+     * @return integer
+     */
+    protected function count(string $key, int $datetime): int
+    {
+        if (0 == $this->limit) { // No limit no point for counting
+            return 0;
+        }
+
+        if (! \is_null($this->count)) { // Get fetched result
+            return $this->count;
+        }
+
+        $count = $this->redis->get(self::NAMESPACE . ':'. $key .':'. $datetime);
+        if (!$count) {
+            $this->count = 0;
+        } else {
+            $this->count = (int) $count;
+        }
+
+        return $this->count;
+    }
+
+    /**
+     * @param  string  $key
+     * @param  int  $datetime
+     * @return void
+     *
+     */
+    protected function hit(string $key, int $datetime): void
+    {
+        if (0 == $this->limit) { // No limit no point for counting
+            return;
+        }
+
+        $count = $this->redis->get(self::NAMESPACE . ':'. $key .':'. $datetime);
```

Comment:
On our benchmarks it would be great to test performance with and without the `:`. We suspect they might be adding some cost to performance at scale.

Reply:
Maybe also add some auto expiry to keys so we don't need to deal with cleanups?

Reply:
or we already do cleanups?

Reply:
As for cleanup, Similar to database adapter we have a cleanup method that accepts a date as `timestamp` string and deletes all the logs up to that date.
`timestamp` is required as string to keep the function definition similar to database adapter (where database adapter accepts datetime in string)

Reply:
were we able to test how it performs with many keys? My fear is that we'll have very slow full scans if the list of keys will get big.

Reply:
I did run the benchmarks and did not understand the result. I will try to run manual tests with huge number of keys